### PR TITLE
Fix ctrl-d/u issues with scroll_beyond_last_line off

### LINKED ISSUE #15356: Vim: ctrl-d and ctrl-u bugs when used with `scroll_beyond_last_line` off
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

There are three related buggy behaviors. I'd be happy to split these into separate issues if you'd prefer that.

All of these are with the setting `"scroll_beyond_last_line": "off"`, which matches the behavior of neovim.

1. `ctrl-d` can't reach the bottom of the buffer.
2. If the cursor is already further down than the point it could otherwise reach via `ctrl-d`, it will actually go up.
3. When the cursor is on the last line, `ctrl-u` will scroll up, but the cursor won't move with it. It stays on the last line.

### Environment

Zed: v0.145.1 (Zed)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

Video showing the first two bugs:

https://github.com/user-attachments/assets/6d0fddd1-3c10-4e53-ba58-13c995708deb

Video showing the third bug:

https://github.com/user-attachments/assets/f0ba772d-6798-4bcb-8594-27cf5548f32f

The same commands in neovim:

https://github.com/user-attachments/assets/52ec847d-dac8-4c8c-93d7-b3c6eed907bb

### If applicable, attach your Zed.log file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by ksweetie**
Closes #15356

Release Notes:

- vim: Fixed issues with `ctrl-d`/`ctrl-u` when `scroll_beyond_last_line` is set to `off` ([#15356](https://github.com/zed-industries/zed/issues/15356)).

https://github.com/user-attachments/assets/d3166393-4a4e-4195-9db6-3ff1d4aeec78

### CODE REVIEW

**Comment by ksweetie** [comment]
@cla-bot check

File: crates/editor/src/editor.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ mod blame_entry_tooltip;
 mod blink_manager;
 mod debounced_delay;
 pub mod display_map;
-mod editor_settings;
+pub mod editor_settings;
```
Comment (ksweetie) [inline crates/editor/src/editor.rs:20]: I'm guessing there's a better way to do this, but my Rust is very basic. I just wanted to use the `ScrollBeyondLastLine` enum in a test below.

File: crates/vim/src/normal/scroll.rs [hunk 2]
```diff
@@ -91,12 +91,13 @@ fn scroll_editor(
         s.move_with(|map, selection| {
             let mut head = selection.head();
             let top = top_anchor.to_display_point(map);
+            let starting_column = head.column();
```
Comment (ksweetie) [inline crates/vim/src/normal/scroll.rs:94]: I'm storing the starting column due to (I think) a bug in `map.clip_point`, which will move the cursor to the right when passed a display point that is past the last line, even if you provide `Bias::Left`.

There's almost surely a better way to fix this root issue, but using the starting_column after `clip_point` ensures that we attempt to go return to the column we started on, which more closely matches neovim behavior. Neovim's behavior is actually more complex than that, but I think that would be a can of worms for this PR.

This fix isn't necessary for the issue I originally opened. However, when I was writing the test for the original issue, I was running into a difference in behavior between the column that the cursor ended up on between neovim and zed. I can revert this change, but I would likely need to add a hack into the new test to re-adjust the column.

File: crates/editor/src/editor.rs (hunk 1 above)
Comment (maxdeviant) [reply crates/editor/src/editor.rs]: We already re-export some members of `editor_settings`, so we can add `ScrollBeyondLastLine` to the list: b2210a11b6

**Comment by ksweetie** [comment]
@ConradIrwin Hey, fellow Boulderite 😄 Do you have any thoughts on this PR? It looks like there's a merge conflict now. I can fix that if you're on board with the general direction of this change. Let me know what you think. Thanks!

**Comment by ConradIrwin** [comment]
Thank you for taking this on, and sorry for the glacial reply! You managed to hit send just before I was offline for a few weeks.

The changes make sense to me, and I'm happy to merge it. As you rebase it'd be good to share the `fn sample_text()` so we only have one copy of it, but thank you for adding the test!

Also it'd be good to mention `scroll_beyond_last_line` in docs/src/vim.md under "settings you might also like".

**Comment by ksweetie** [comment]
@ConradIrwin No worries, I was starting to get worried that my poor Rust skills scared you off 😄 

I appreciate the feedback. I made those changes. Let me know if I can do anything else for this one.

**Comment by ConradIrwin** [comment]
@ksweetie I didn't notice those, but clippy did :D. Mind fixing that, then I'll merge.

**Comment by ksweetie** [comment]
@ConradIrwin Oops, I think that was part of the merge conflict that I didn't do a great job resolving. I think the fix I just pushed should work.

**Pull request merged**